sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala

```diff
@@ -1106,6 +1107,54 @@ class StreamingQuerySuite extends StreamTest with BeforeAndAfter with Logging wi
     }
   }
 
+  test("union in streaming query of append mode without watermark") {
+    val inputData1 = MemoryStream[Int]
+    val inputData2 = MemoryStream[Int]
+    withTempView("s1", "s2") {
+      inputData1.toDF().createOrReplaceTempView("s1")
+      inputData2.toDF().createOrReplaceTempView("s2")
+      val unioned = spark.sql(
+        "select s1.value from s1 union select s2.value from s2")
+      checkExceptionMessage(unioned)
+    }
+  }
+
+  test("distinct in streaming query of append mode without watermark") {
+    val inputData = MemoryStream[Int]
+    withTempView("deduptest") {
+      inputData.toDF().toDF("value").createOrReplaceTempView("deduptest")
+      val distinct = spark.sql("select distinct value from deduptest")
+      checkExceptionMessage(distinct)
+    }
+  }
+
+  test("distinct in streaming query of complete mode") {
+    val inputData = MemoryStream[Int]
+    withTempView("deduptest") {
+      inputData.toDF().toDF("value").createOrReplaceTempView("deduptest")
+      val distinct = spark.sql("select distinct value from deduptest")
+
+      testStream(distinct, Complete)(
+        AddData(inputData, 1, 2, 3, 3, 4),
+        CheckAnswer(Row(1), Row(2), Row(3), Row(4))
```

Comment:
The behavior is a bit confusing to me actually. What result would we expect if there's another batch `(3, 3, 4, 5)`? 
(I'm not asking which values Spark are providing. I'm asking which values are "reasonable" and "clear".)

Reply:
if there's another batch `(3, 3, 4, 5)`, the result becomes `1,2,3,4,5`, which looks reasonable in complete mode.

Reply:
```
What result would we expect if there's another batch (3, 3, 4, 5)?
```
It should be `1,2,3,4,5`.
I only add one batch here just want to show the distinct in complete can pass the analyzer. Because without this change, we'll get an exception of `Complete output mode not supported when there are no streaming aggregations on streaming DataFrames/Datasets;`.
Let me add this in the test, maybe it's more clear to explain the behavior of `distinct` for multiple batches. The distinct operation should both distinct within a single batch and cross multi-batches.

Is this make sense to you?
```
testStream(distinct, Complete)(
    AddData(inputData, 1, 2, 3, 3, 4),
    CheckAnswer(Row(1), Row(2), Row(3), Row(4)),
    AddData(inputData, 3, 3, 4, 5),
    CheckAnswer(Row(1), Row(2), Row(3), Row(4), Row(5))
)
```

Reply:
If we define the behavior of `distinct` like this, then `dropDuplicates` is no longer alias of `distinct`. `dropDuplicates` ensures the row is produced "only once" regardless of the output mode. (Like the stream-stream join does.)

Which one is incorrect? Is dropDuplicates a special operator which isn't bound to aggregation (hence result table doesn't apply here) while it still plays as aggregator behind the scene?

Reply:
This is the reason I don't like complete mode - for me it has been very confusing, and force me to think which operators support this mode. Not all stateful operators support the mode.

If we miss to restrict the operators which don't work with complete mode, we'll end up with losing all outputs in previous batches, especially we now clearly require sink to truncate the old output for complete mode.

Reply:
Note, `dropDuplicates is an alias of distinct`, this refers to the DataSet API dropDuplicates is an alias of distinct. What we are fixing here is the SQL distinct. The Dataset APIs do not keep the same with SQL ones, I think that's known issues(e.g SQL union vs Dataset.union). It's a good discussion of the compatibility between Dataset and SQL API, but I think it should be another topic and we can fix it in other PR if necessary.

Reply:
Yea I agree. It's better to make DataFrame and SQL APIs consistent, but it's already an issue and we should not block this PR.

Reply:
I'd be more puzzled we differentiate the behavior of streaming query via DF vs SQL. (I never use SQL for SS query though.) Do we think end users can make correct assumption on differentiating twos? Otherwise do we have any guide or note on the doc?

After the fix, end users have to infer union & distinct triggers grouping aggregation and hence result table, even without explicit group by. I agree this patch fixes the thing which didn't work, but brings confusion on reasoning about "implicit behavior" which depends on the implementation details for Spark. Probably we should make it clear "grouped aggregation" includes implicit aggregation in the guide doc.

And I'm wondering whether we track a list of "known issues" and any plan to fix them. The problems on update/complete mode are already pointed out during couple of Spark 3.0.0 related issues, and I don't see any plan on addressing it. I even proposed to replace complete mode with viable alternatives but brought the debates and no productive discussion happened. I don't think being conservative means we live with known problems.

Reply:
I think the most valuable part of this patch is to improve the error message from `java.util.NoSuchElementException: None.get` to requesting watermark. Supporting complete mode doesn't really matter. If you think it's confusing, feel free to open a PR to ban it.

I agree that there are still problems around the output mode, but this PR itself is nothing wrong. `Distinct` is planned as `Aggregate` and we must treat it as `Aggregate` in the streaming checker.

Reply:
http://spark.apache.org/docs/latest/structured-streaming-programming-guide.html#unsupported-operations

It has been clear on the structured streaming guide doc - `distinct` is unsupported. I'm not sure this patch also enables the case from DataFrame API. If it fixes the DF case, worth to update the doc. If not, this will be another case of "known inconsistency".

So I think we have been trying to restrict the operations which don't make sense or error prone in streaming Datasets (unbounded input) - that has been working with DataFrame API (I think that's the reason we have "specific" methods like union which doesn't work like SQL-like union and dropDuplicates which ignores output mode), and is broken with allowing SQL statements in streaming Datasets.

@tdas @zsxwing @jose-torres @brkyvz Would like to hear your thought about this as well, as probably I'm missing something here on the purpose.

Reply:
Are you suggesting to ban `Distinct` in SS completely? I think it's fine too, as long as we don't give confusing error messages.

Reply:
What I am suggesting is that waiting and hearing the operations we have been restricted on SS with the reasons, and if the reasons make sense then ban them even with SQL statements. Not only distinct.

Reply:
As an alternative I added some note on SS guide doc. #29461

I'm not sure it is enough to let us free to not complained by improper usages, so I just marked the PR as draft. I think it's better to collect the voices on this.

Reply:
Agree to have notes on doc and comments. 
This fix is only for SQL DISTINCT clause. `It has been clear on the structured streaming guide doc - distinct is unsupported.` This is also referred to the Dataset API `distinct`, not the SQL DISTINCT clause.

Reply:
Also, the different behavior between SQL UNION and Dataset.union I mentioned [here](https://github.com/apache/spark/pull/29256#discussion_r466966080) is not only for Streaming. The Dataset.union is equivalent to `UNION ALL` in SQL, and SQL UNION hasn't corresponded API in Dataset yet.

Reply:
I think the right way is unifying the behavior of SQL distinct and Dataset.dropDuplicates(). That is to say, the logic node of `Distinct` should be replaced to `Deduplicate` in streaming. It should easily work if we did it at the beginning. But now we need more work to finish the behavior unification. Because the fix will also change the current state store format, so the fix depends on state store versioning.

Reply:
> It has been clear on the structured streaming guide doc - distinct is unsupported. This is also referred to the Dataset API distinct, not the SQL DISTINCT clause.

Are you sure that can be interpreted as same for end users? Is there any mention of Dataset API vs SQL clause in the statement?

> Also, the different behavior between SQL UNION and Dataset.union I mentioned here is not only for Streaming. The Dataset.union is equivalent to UNION ALL in SQL, and SQL UNION hasn't corresponded API in Dataset yet.

That's clearly described in the doc. The thing is that SQL UNION cannot be done in streaming, as according to the doc, SQL UNION is Dataset.union + distinct but distinct is not supported. So this is another one being enabled as side effects. You can't do that with Dataset API.

Reply:
```
Are you sure that can be interpreted as same for end users? Is there any mention of Dataset API vs SQL clause in the statement?
```
Let me clarify more. I'm not sure so I agree we should have more documents. What I'm suggesting is this document actually is not Streaming specific, it's the inconsistency between SQL and Dataset API.
For the structured streaming guide doc, I think the `distinct` it only refers to Dataset.distinct, not the SQL DISTINCT clause.

```
That's clearly described in the doc. The thing is that SQL UNION cannot be done in streaming, as according to the doc, SQL UNION is Dataset.union + distinct but distinct is not supported.
```
Yep, so I think we are on the same page. That's not clear enough right? The best way is to directly tell the end user SQL union not support in Streaming, because it's relay on distinct. The root cause still the inconsistency between SQL and Dataset.

```
 So this is another one being enabled as side effects. You can't do that with Dataset API.
```
That's why my suggestion is making the distinct and drop duplicates have the same behavior.

Reply:
Sorry I was really busy with handling stuffs in my plate.

I'm less concerned about the difference between SQL distinct vs Dataset API distinct for append/update mode. There's still difference between twos about when the output is available, but the output is available only once for both side.
(It would be better to guide about the effects for the SQL statement applied on streaming dataset. I'll turn #29461 to be 'ready to review'.)

What I'm concerned about is the complete mode, but I'd admit my concern for the complete mode is not limited to this PR. I'll try to make it be correct in other place, instead of this PR.